src/platform/nrfconnect/FactoryDataProvider.h

```diff
@@ -21,16 +21,37 @@
 #include <platform/CommissionableDataProvider.h>
 #include <platform/DeviceInstanceInfoProvider.h>
 
-#include <zephyr/drivers/flash.h>
 #include <fprotect.h>
 #include <pm_config.h>
 #include <system/SystemError.h>
+#include <zephyr/drivers/flash.h>
 
 #include "FactoryDataParser.h"
 
 namespace chip {
 namespace DeviceLayer {
 
+// These two helpers allows to get the factory data memory block which shall be protected with fprotect, so that:
+// 1) it is aligned to the multiple of CONFIG_FPROTECT_BLOCK_SIZE (which differs between nRF families)
+// 2) it does not exceed the settings partition start address
+// Note that this block can overlap with app partition but this is not a problem since we do not aim to modify
+// the application code at runtime anyway.
+constexpr size_t FactoryDataBlockBegin()
+{
+    // calculate the nearest multiple of CONFIG_FPROTECT_BLOCK_SIZE smaller than PM_FACTORY_DATA_ADDRESS
+    return PM_FACTORY_DATA_ADDRESS & (-CONFIG_FPROTECT_BLOCK_SIZE);
+}
+
+constexpr size_t FactoryDataBlockSize()
+{
+    // calculate the address which includes the PM_FACTORY_DATA_SIZE but give a chance to add another offset
+    // to obtain the multiple of CONFIG_FPROTECT_BLOCK_SIZE (but make sure we do not overlap with settings)
+    constexpr size_t kFActoryDataBlockEnd =
+        (PM_FACTORY_DATA_ADDRESS + PM_FACTORY_DATA_SIZE + PM_FACTORY_DATA_SIZE - 1) & (-CONFIG_FPROTECT_BLOCK_SIZE);
```

Comment:
```suggestion
        (PM_FACTORY_DATA_ADDRESS + PM_FACTORY_DATA_SIZE + CONFIG_FPROTECT_BLOCK_SIZE - 1) & (-CONFIG_FPROTECT_BLOCK_SIZE);
```
sorry, I made a mistake yesterday.  `(x + BLOCK_SIZE) & (-BLOCK_SIZE)` is the way to round `x` up to the nearest multiple of `BLOCK_SIZE`.

Reply:
discussed f2f